### PR TITLE
Add request date validation

### PR DESCRIPTION
**Pull request by bi-shun**
Handling dates in Signature Version 4
Ref: https://docs.aws.amazon.com/general/latest/gr/sigv4-date-handling.html

Signed-off-by: BI SHUN KE <aionshun@livemail.tw>

### CODE REVIEW

File: cmd/signature-v4.go [hunk 1]
```diff
@@ -280,6 +280,11 @@ func doesPresignedSignatureMatch(hashedPayload string, r *http.Request, region s
 	// Get the encoded query.
 	encodedQuery := query.Encode()
 
+	// Verify if credential scope date is same request date
+	// - https://docs.aws.amazon.com/general/latest/gr/sigv4-date-handling.html
+	if pSignValues.Date.Format(yyyymmdd) != t.Format(yyyymmdd) {
+		return ErrAuthorizationDateHeaderMalformed
+	}
 	// Verify if date query is same.
 	if req.URL.Query().Get(xhttp.AmzDate) != query.Get(xhttp.AmzDate) {
```
Comment (harshavardhana) [inline cmd/signature-v4.go:289]: @bi-shun we already have this here

File: cmd/signature-v4.go [hunk 2]
```diff
@@ -368,6 +373,12 @@ func doesSignatureMatch(hashedPayload string, r *http.Request, region string, st
 		return ErrMalformedDate
 	}
 
+	// Verify if credential scope date is same request date
+	// - https://docs.aws.amazon.com/general/latest/gr/sigv4-date-handling.html
+	if signV4Values.Credential.scope.date.Format(yyyymmdd) != t.Format(yyyymmdd) {
```
Comment (harshavardhana) [inline cmd/signature-v4.go:378]: This looks correct

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#7105a85c-e891-4ced-ba4d-36726d0c26f8) |
| mint-gateway-nas.sh | :x: [more...](#b58d0af6-bfce-4545-a460-c684a3aaa1e3) |
<a name="b58d0af6-bfce-4545-a460-c684a3aaa1e3"></a>
### 9421-a08c3aa/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev6.minio.io:32587
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 3d436366e796:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 0 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 41 seconds
(4/15) Running aws-sdk-ruby tests ... done in 2 seconds
(5/15) Running awscli tests ... done in 2 minutes and 10 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 27 seconds
(8/15) Running minio-dotnet tests ... done in 30 seconds
(9/15) Running minio-go tests ... FAILED in 28 seconds
{}

Executed 8 out of 15 tests successfully.
```
<a name="7105a85c-e891-4ced-ba4d-36726d0c26f8"></a>
### 9421-a08c3aa/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev3.minio.io:30904
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 65001d3cbd0d:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 7 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 1 minutes and 4 seconds
(4/15) Running aws-sdk-ruby tests ... done in 15 seconds
(5/15) Running awscli tests ... FAILED in 55 seconds
{
  "name": "awscli",
  "duration": 4092,
  "function": "aws --endpoint-url http://minio-dev3.minio.io:30904 s3api copy-object --bucket awscli-mint-test-bucket-12937 --storage-class REDUCED_REDUNDANCY --key datafile-1-kB-copy --copy-source awscli-mint-test-bucket-12937/datafile-1-kB\n",
  "status": "FAIL",
  "error": "An error occurred (NotImplemented) when calling the CopyObject operation: A header you provided implies functionality that is not implemented"
}

Executed 4 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

File: cmd/signature-v4.go (hunk 1 above)
Comment (bi-shun) [reply cmd/signature-v4.go:289]: > @bi-shun we already have this here

Thank you for your comment. 
As far as I know, credential date is equal to same x-amz-date.
Now, I correct it using credential date to compare.

**Comment by harshavardhana** [comment]
Please re-submit the PR with fixes, closing this PR until then

**Pull request closed without merging**